pkg/cmd/server/api/validation/master.go

```diff
@@ -0,0 +1,169 @@
+	allErrs = append(allErrs, ValidateKubeConfig(config.MasterClients.OpenShiftLoopbackKubeConfig, "openShiftLoopbackKubeConfig").Prefix("masterClients")...)
+	allErrs = append(allErrs, ValidateKubeConfig(config.MasterClients.KubernetesKubeConfig, "kubernetesKubeConfig").Prefix("masterClients")...)
+
+	allErrs = append(allErrs, ValidatePolicyConfig(config.PolicyConfig).Prefix("policyConfig")...)
+
+	allErrs = append(allErrs, ValidateServingInfo(config.ServingInfo).Prefix("servingInfo")...)
+
+	return allErrs
+}
+
+func ValidateAssetConfig(config *api.AssetConfig) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	allErrs = append(allErrs, ValidateServingInfo(config.ServingInfo).Prefix("servingInfo")...)
+
+	urlObj, urlErrs := ValidateURL(config.PublicURL, "publicURL")
+	if len(urlErrs) > 0 {
+		allErrs = append(allErrs, urlErrs...)
+	}
+	if urlObj != nil {
+		if !strings.HasSuffix(urlObj.Path, "/") {
+			allErrs = append(allErrs, errs.NewFieldInvalid("publicURL", config.PublicURL, "must have a trailing slash in path"))
+		}
+	}
+
+	return allErrs
+}
+
+func ValidateEtcdConnectionInfo(config api.EtcdConnectionInfo) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if len(config.URLs) == 0 {
+		allErrs = append(allErrs, errs.NewFieldRequired("urls"))
+	}
+	for i, u := range config.URLs {
+		_, urlErrs := ValidateURL(u, fmt.Sprintf("urls[%d]", i))
+		if len(urlErrs) > 0 {
+			allErrs = append(allErrs, urlErrs...)
+		}
+	}
+
+	if len(config.CA) > 0 {
+		allErrs = append(allErrs, ValidateFile(config.CA, "ca")...)
+	}
+	allErrs = append(allErrs, ValidateCertInfo(config.ClientCert, false)...)
+
+	return allErrs
+}
+
+func ValidateEtcdConfig(config *api.EtcdConfig) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	allErrs = append(allErrs, ValidateServingInfo(config.ServingInfo).Prefix("servingInfo")...)
+
+	if len(config.StorageDir) == 0 {
+		allErrs = append(allErrs, errs.NewFieldRequired("storageDirectory"))
+	}
+
+	return allErrs
+}
+
+func ValidateImageConfig(config api.ImageConfig) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if len(config.Format) == 0 {
+		allErrs = append(allErrs, errs.NewFieldRequired("format"))
+	}
+
+	return allErrs
+}
+
+func ValidateKubeletConnectionInfo(config api.KubeletConnectionInfo) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if config.Port == 0 {
+		allErrs = append(allErrs, errs.NewFieldRequired("port"))
+	}
+
+	if len(config.CA) > 0 {
+		allErrs = append(allErrs, ValidateFile(config.CA, "ca")...)
+	}
+	allErrs = append(allErrs, ValidateCertInfo(config.ClientCert, false)...)
+
+	return allErrs
+}
+
+func ValidateKubernetesMasterConfig(config *api.KubernetesMasterConfig) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if len(config.MasterIP) > 0 {
+		allErrs = append(allErrs, ValidateSpecifiedIP(config.MasterIP, "masterIP")...)
```

Comment:
Not directly related, but you moved it so it caught my eye.  In the interest of fully specifying, should we simply fail on empty MasterIP?  I'm for it.